nativelib/src/main/scala/scala/scalanative/unsafe/package.scala

```diff
@@ -112,6 +112,11 @@ package object unsafe extends unsafe.UnsafePackageCompat {
    */
   final class extern extends scala.annotation.StaticAnnotation
 
+  /** An annotation that is used to mark methods that contain externally-defined
+   *  and potentially blocking methods
+   */
+  final class blocking extends scala.annotation.StaticAnnotation
```

Comment:
I'm experimenting with StopTheWorld events for the GC. In general it was previously hard to make a good balance in terms of suspending the threads in a safe manner. Using signals on it's own is quite unsafe - we can stop in the middle of sequences of operations that should be treated as atomic. In the past it lead to some undefined behaviours. 
Because of that, every call to native/extern function was treated as potentially blocking - before calling the method we would flip the flag in GC mutator thread, and would store the current top address of the stack and since recently we would dump the registers. Both of this information is needed to correctly mark the objects. This however introduced a lot of new problems, because calls to non blocking functions, eg. some constant value was also requiring this overhead. 

By introducing the blocking attribute we should be able to introduce some kind of balance, even though the user experience might get a bit worse, because if you add the binds you should remember about annotation for possibly blocking calls

Reply:
Thanks, that is a great explanation! This does seem like a good optimization.

> because if you add the binds you should remember about annotation for possibly blocking calls

In Cats Effect we must wrap blocking calls with `IO.blocking(...)` anyway, so we are used to this :) actually I was wondering if we could use this annotation in Cats Effect somehow.